core/crypto/bls_provider/bls_provider_impl.cpp

```diff
@@ -0,0 +1,113 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "crypto/bls_provider/bls_provider_impl.hpp"
+
+extern "C" {
+#include "crypto/bls_provider/libbls_signatures.h"
+}
+
+#include <gsl/gsl_util>
+
+namespace fc::crypto::bls {
+  outcome::result<KeyPair> BlsProviderImpl::GenerateKeyPair() const {
+    KeyPair key_pair{};
+    PrivateKeyGenerateResponse *private_key_response = private_key_generate();
+    if (private_key_response == nullptr) {
+      return Errors::KeyPairGenerationFailed;
+    }
+    auto free_private_response = gsl::finally([private_key_response]() {
+      destroy_private_key_generate_response(private_key_response);
+    });
+    std::move(std::begin(private_key_response->private_key),
+              std::end(private_key_response->private_key),
+              key_pair.private_key.begin());
+    OUTCOME_TRY(public_key, DerivePublicKey(key_pair.private_key));
+    key_pair.public_key = public_key;
+    return key_pair;
+  }
+
+  outcome::result<PublicKey> BlsProviderImpl::DerivePublicKey(
+      const PrivateKey &key) const {
+    PublicKey public_key;
+    PrivateKeyPublicKeyResponse *response = private_key_public_key(key.data());
+    if (response == nullptr) {
+      return Errors::InvalidPrivateKey;
+    }
+    auto free_response = gsl::finally(
+        [response]() { destroy_private_key_public_key_response(response); });
+    std::move(std::begin(response->public_key),
+              std::end(response->public_key),
+              public_key.begin());
+    return public_key;
+  }
+
+  outcome::result<Signature> BlsProviderImpl::Sign(
+      gsl::span<uint8_t> message, const PrivateKey &key) const {
+    Signature signature;
+    PrivateKeySignResponse *response =
+        private_key_sign(key.data(), message.data(), message.size());
+    if (response == nullptr) {
+      return Errors::SignatureGenerationFailed;
+    }
+    auto free_response = gsl::finally(
+        [response]() { destroy_private_key_sign_response(response); });
+    std::move(std::begin(response->signature),
+              std::end(response->signature),
+              signature.begin());
+    return signature;
+  }
+
+  outcome::result<bool> BlsProviderImpl::VerifySignature(
+      gsl::span<uint8_t> message,
+      const Signature &signature,
+      const PublicKey &key) const {
+    OUTCOME_TRY(digest, GenerateHash(message));
+    if (verify(signature.data(),
+               digest.data(),
+               digest.size(),
+               key.data(),
+               key.size())
+        == 1) {
```

Comment:
Minor issue: usually rule is: 0 is false, everything else is true. I think here it is also the case
```suggestion
        != 0) {
```